zrml/prediction-markets/src/lib.rs

```diff
@@ -1046,20 +1122,30 @@ mod pallet {
     #[pallet::hooks]
     impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
         fn on_initialize(now: T::BlockNumber) -> Weight {
-            let mut total_weight: Weight = 0;
-            let rslt = Self::resolution_manager(now, |market_id, market| {
-                let weight = Self::on_resolution(market_id, market)?;
-                total_weight = total_weight.saturating_add(weight);
-                Ok(())
-            });
-            with_transaction(|| match rslt {
-                Err(err) => {
-                    Self::deposit_event(Event::BadOnInitialize);
-                    log::error!("Block {:?} was not initialized. Error: {:?}", now, err);
-                    TransactionOutcome::Rollback(())
+            let mut total_weight: Weight =
+                Self::process_subsidy_collecting_markets(now, T::MarketCommons::now());
+
+            let mut do_resolution = || {
+                Self::resolution_manager(now, |market_id, market| {
+                    let weight = Self::on_resolution(market_id, market)?;
+                    total_weight = total_weight.saturating_add(weight);
+                    Ok(())
+                })
+            };
```

Comment:
Not sure why this method needs to be wrapped inside a closure. You can move the whole `Self::resolution_manager(...)` inside `with_transaction`.